File: Classes/Layers/DrawLayer.m
/*
DrawLayer.m
Draw

Copyright © 2021, AJ Raftis and AJRFoundation authors
All rights reserved.

Redistribution and use in source and binary forms, with or without modification,
are permitted provided that the following conditions are met:

* Redistributions of source code must retain the above copyright notice, this 
  list of conditions and the following disclaimer.
* Redistributions in binary form must reproduce the above copyright notice, 
  this list of conditions and the following disclaimer in the documentation 
  and/or other materials provided with the distribution.
* Neither the name of Draw nor the names of its contributors may be 
  used to endorse or promote products derived from this software without 
  specific prior written permission.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED 
WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE 
DISCLAIMED. IN NO EVENT SHALL AJ RAFTIS BE LIABLE FOR ANY DIRECT, INDIRECT, 
INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR 
PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF 
LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE
OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF 
ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
*/

#import "DrawLayer.h"

#import "DrawDocument.h"
#import <Draw/Draw-Swift.h>
#import "AJRXMLCoder-DrawExtensions.h"

@interface DrawDocument (PrivateLayer)

- (void)_resetLayerPopUpButton;
- (void)replaceNameOnLayerFromName:(NSString *)oldName toName:(NSString *)newName;

@end

@implementation DrawLayer

- (id)initWithName:(NSString *)aName document:(DrawDocument *)aDocumentView {
    if ((self = [super init])) {
        // Set should be done before setDrawView, since once we do that we'll have an undo manager.
        _variableStore = [[AJRStore alloc] init];
        [self setName:aName];
        [self setVisible:YES];
        [self setPrintable:YES];
        [self setLocked:NO];
        [self setDocument:aDocumentView];
    }
    return self;
}

#pragma mark - Properties

- (void)setName:(NSString *)aName {
    if (_name != aName && ![_name isEqualToString:aName]) {
        [[_document prepareWithInvocationTarget:_document] replaceNameOnLayerFromName:aName toName:_name];
        _name = aName;
        [_document _resetLayerPopUpButton];
    }
}

- (void)setLocked:(BOOL)flag {
    if (_locked != flag) {
        [[_document prepareWithInvocationTarget:self] restoreFromSnapshot:self.snapshot];
        _locked = flag;
        [_document setPagesNeedDisplay:YES];
    }
}

- (void)setVisible:(BOOL)flag {
    if (_visible != flag) {
        [[_document prepareWithInvocationTarget:self] restoreFromSnapshot:self.snapshot];
        _visible = flag;
        [_document setPagesNeedDisplay:YES];
    }
}

- (void)setPrintable:(BOOL)flag {
    if (_printable != flag) {
        [[_document prepareWithInvocationTarget:self] restoreFromSnapshot:self.snapshot];
        _printable = flag;
        [_document setPagesNeedDisplay:YES];
    }
}

- (void)setDocument:(DrawDocument *)document {
    // Make sure our variables points to our document.
    [_variableStore enumerate:^(NSString *name, id <AJREvaluation> value, BOOL *stop) {
        DrawVariable *variable = AJRObjectIfKindOfClass(value, DrawVariable);
        if (variable != nil) {
            variable.document = document;
            variable.layer = self;
        }
    }];
}

#pragma mark - Snapshotting

- (NSDictionary *)snapshot {
    return @{@"name":_name, @"locked":@(_locked), @"visible":@(_visible), @"printable":@(_printable)};
}

- (void)restoreFromSnapshot:(NSDictionary *)snapshot {
    [self setName:snapshot[@"name"]];
    [self setVisible:[snapshot[@"visible"] boolValue]];
    [self setPrintable:[snapshot[@"printable"] boolValue]];
    [self setLocked:[snapshot[@"locked"] boolValue]];
}

#pragma mark - AJRXMLCoding

+ (NSString *)ajr_nameForXMLArchiving {
    return @"layer";
}

- (void)decodeWithXMLCoder:(AJRXMLCoder *)coder {
    [coder decodeObjectForKey:@"name" setter:^(id  _Nullable object) {
        self->_name = object;
    }];
    [coder decodeBoolForKey:@"locked" setter:^(BOOL value) {
        self->_locked = value;
    }];
    [coder decodeBoolForKey:@"visible" setter:^(BOOL value) {
        self->_visible = value;
    }];
    [coder decodeBoolForKey:@"printable" setter:^(BOOL value) {
        self->_printable = value;
    }];
    [coder decodeObjectForKey:@"variableStore" setter:^(id object) {
        if (object == nil) {
            self->_variableStore = [[AJRStore alloc] init];
        } else {
            self->_variableStore = object;
        }
    }];
}

- (void)encodeWithXMLCoder:(AJRXMLCoder *)coder {
    [coder encodeString:_name forKey:@"name"];
    [coder encodeBool:_locked forKey:@"locked"];
    [coder encodeBool:_visible forKey:@"visible"];
    [coder encodeBool:_printable forKey:@"printable"];
    if (_variableStore.count > 0) {
        // Let's only encode this if it matters, since it usually won't.
        [coder encodeObject:_variableStore forKey:@"variableStore"];
    }
}

@end

File: Classes/Layers/DrawLayer.swift
/*
 DrawLayer.m
 Draw

 Copyright © 2022, AJ Raftis and Draw authors
 All rights reserved.

 Redistribution and use in source and binary forms, with or without modification,
 are permitted provided that the following conditions are met:

 * Redistributions of source code must retain the above copyright notice, this
   list of conditions and the following disclaimer.
 * Redistributions in binary form must reproduce the above copyright notice,
   this list of conditions and the following disclaimer in the documentation
   and/or other materials provided with the distribution.
 * Neither the name of Draw nor the names of its contributors may be
   used to endorse or promote products derived from this software without
   specific prior written permission.

 THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
 ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
 WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
 DISCLAIMED. IN NO EVENT SHALL AJ RAFTIS BE LIABLE FOR ANY DIRECT, INDIRECT,
 INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
 LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
 PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
 LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE
 OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF
 ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

import AJRInterface

@objcMembers
open class DrawLayer : NSObject, AJRXMLCoding {

    // MARK: - Properties

    open weak var document : DrawDocument? {
        didSet {
            for variable in variableStore {
                if let variable = variable as? DrawVariable {
                    variable.document = document
                    variable.layer = self
                }
            }
        }
    }
    open var name = "" {
        willSet {
            let captured = self.name
            document?.registerUndo(target: self, handler: { target in
                target.name = captured
            })
        }
        didSet {
            document?.noteLayersChanged()
        }
    }
    open var locked = false {
        willSet {
            let captured = self.locked
            document?.registerUndo(target: self, handler: { target in
                target.locked = captured
            })
        }
        didSet {
            document?.pagesNeedDisplay = true
        }
    }
    open var visible = true {
        willSet {
            let captured = self.visible
            document?.registerUndo(target: self, handler: { target in
                target.locked = captured
            })
        }
        didSet {
            document?.pagesNeedDisplay = true
        }
    }
    open var printable = true {
        willSet {
            let captured = self.printable
            document?.registerUndo(target: self, handler: { target in
                target.locked = captured
            })
        }
        didSet {
            document?.pagesNeedDisplay = true
        }
    }
    open var variableStore = AJRStore()

    // MARK: - Creation

    required public override init() {
        // Only used by archiving
        super.init()
    }

    @objc(initWithName:document:)
    public init(name: String, document: DrawDocument) {
        // Set should be done before setDrawView, since once we do that we'll have an undo manager.
        self.name = name
        self.document = document

        super.init()
    }

    // MARK: - Variables

    // TODO: Currently needed for Obj-C compatibility, but the next method is preferred.
    open func addVariablesTo(_ variables : NSMutableArray) {
        for name in variableStore.orderedNames {
            if let variable = variableStore[name] as? AJRVariable {
                if !variables.contains(variable) {
                    variables.add(variable)
                }
            }
        }
    }

    open func addVariables(to variables: inout [AJRVariable]) {
        for name in variableStore.orderedNames {
            if let variable = variableStore[name] as? AJRVariable {
                if !variables.contains(variable) {
                    variables.append(variable)
                }
            }
        }
    }

    open var variables : [AJRVariable] {
        var variables = [AJRVariable]()
        addVariables(to: &variables)
        return variables
    }

    // MARK: - Snapshotting

    // TODO: Might not want these method anymore.

    open var snapshot : [String:Any] {
        return ["name":name, "locked":locked, "visible":visible, "printable":printable]
    }

    @objc(restoreFromSnapshot:)
    open func restore(from snapshot: [String:Any]) {
        name = snapshot["name", "Unnamed"]
        visible = snapshot["visible", true]
        printable = snapshot["printable", true]
        locked = snapshot["locked", false]
    }

    // MARK: - AJRXMLCoding

    open override class var ajr_nameForXMLArchiving: String {
        return "layer"
    }

    public func decode(with coder: AJRXMLCoder) {
        coder.decodeString(forKey: "name") { name in
            self.name = name
        }
        coder.decodeBool(forKey: "locked") { value in
            self.locked = value
        }
        coder.decodeBool(forKey: "visible") { value in
            self.visible = value
        }
        coder.decodeBool(forKey: "printable") { value in
            self.printable = value
        }
        coder.decodeObject(forKey: "variableStore") { store in
            if let store = store as? AJRStore {
                self.variableStore = store
            } else {
                // Created during init, so we don't need to.
            }
        }
    }

    public func encode(with coder: AJRXMLCoder) {
        coder.encode(name, forKey: "name")
        coder.encode(locked, forKey: "locked")
        coder.encode(visible, forKey: "visible")
        coder.encode(printable, forKey: "printable")
        if variableStore.count > 0 {
            // Let's only encode this if it matters, since it usually won't.
            coder.encode(variableStore, forKey: "variableStore")
        }
    }

}
